server/src/com/thoughtworks/go/server/scheduling/BuildCauseProducerService.java

```diff
@@ -249,6 +258,52 @@ private void updateChangedRevisions(CaseInsensitiveString pipelineName, BuildCau
         materialChecker.updateChangedRevisions(pipelineName, buildCause);
     }
 
+    void pruneRevisionsIfRequired(BuildType buildType, PipelineConfig pipelineConfig, BuildCause buildCause) {
```

Comment:
Also, isn't there some code somewhere in this area, which removes a build from queue and reschedules it, when a new commit happens, and the build hasn't been picked up yet? That'll need to change too, right? It shouldn't remove the old build from queue any more.